ADR/0039-workspace-deprecation.md

```diff
@@ -0,0 +1,98 @@
+# 39. Workspace Deprecation
+
+Some of the text was taken from - https://github.com/konflux-ci/architecture/pull/187
+
+## Status
+
+WiP
+
+## Context
+
+The purpose of this ADR is to revisit the workspace concept, understand the purpose
+of it, and offer an alternative implementation based on native Kubernetes APIs
+and other successful cloud native open source projects.
+
+We inherited the "workspace" concept from our early days when we expected to be integrated with the [KCP] project. Long ago (over a year at the time of this writing), we decided to remove [KCP] from the architecture of the Konflux project.
+
+In order to support multi-cluster deployment, [kubesaw] was used instead of KCP.
+
+Today, Workspaces are an important concept in the Konflux user experience. The user onboards into a workspace of their own. A team member of theirs may add them to a team workspace. As a platform administrator, if you hear about a user having trouble, you'll ask them what workspace they're working in.
+
+A Workspace is an abstraction of a namespace. Each workspace has a single namespace
```

Comment:
No additional resources are needed for running Konflux pipelines. The service account used to run the Pipeline's pods is created by Tekton.

Reply:
Quotas is one now, tomorrow it could be a secret to connect to a database or vault. Or it could be a volume for share data. I'm think more in the future to make Konflux look valuable beyond our pipeline and release features

Reply:
This is the list of resources which is part of a workspace **today**:
- Resource Quotas & LimitRange
- Network Policies
- RBAC & Service Accounts
- ConfigMap

So it's already more than just a namespace today. And as @dperaza4dustbit mentioned it can be even more tomorrow.

Reply:
@alexeykazakov but all of those are not necessary to get a working Konflux for a namespace, this is the reason I treat them as optional.

In this ADR I suggest to use tools such as (but no limited to) ArgoCD and Kyverno to get the resource you mentioned in every tenant namespace for Konflux deployments that needs them.

Reply:
> Quotas is one now, tomorrow it could be a secret to connect to a database or vault. Or it could be a volume for share data. I'm think more in the future to make Konflux look valuable beyond our pipeline and release features

We can discuses solutions when we will have a concrete requirement, as of today Konflux isn't handling deployments and is used only for build, test and release.

Reply:
But you don't have to use the workspace API if you don't want to. Workspace API was Konflux's requirement - it was implemented only for the Konflux needs. It's not used in DevSandbox at all.

Apart from that, there are two things to be supported in KubeSaw, and they seem to be every similar to what you need:

**namespace management via Spaces**
as I showed in this video https://youtu.be/paP2sahtcEc where user provisioning is an implementation/configuration detail of the cluster

**namespace content management via labels/annotations**
We want to support labels/annotations at the level of Namespace resources where we wouldn't manage the creation/deletion of the namespaces, but only the content.
Let's say you create a Namespace via ArgoCD and you want to tell KubeSaw to apply a template there & share it with someone via labels:
```
kind: Namespace
apiVersion: v1
metadata:
  name: your-namespace
  labels:
    toolchain.dev.openshift.com/managed-by-kubesaw: true
    toolchain.dev.openshift.com/template: konflux-large
    toolchain.dev.openshift.com/share-with-admins: gall,david,matous
```
KubeSaw would then apply a set of templates and you (as an admin) won't have to touch the Space CR at all

Reply:
Hypothetically, let's say Konflux directly uses kubernetes-native resources, e.g. namespaces. It then relies on other tooling to manage those resources, e.g. ArgoCD. Konflux and ArgoCD don't know anything about each other. They can be used in conjunction, but don't have to.

Would it be possible to use kubesaw this way? Can't namespaces be created, out of band, via kubesaw?

I asked a similar question during the community meeting, but this is still not clear to me.

Reply:
> Would it be possible to use kubesaw this way? Can't namespaces be created, out of band, via kubesaw?

Yes. Absolutely.

Reply:
If the are no objections I'll close this thread.